File: src/main/java/typeSpecifications/transducerSpecification/TransducerNode.java
package typeSpecifications.transducerSpecification;

import java.util.LinkedList;

// class for specifying a transducer node.
// similar to the Transducer.TransducerNode we have an identifier and booleans for initial or final states.
public class TransducerNode {

    private String identifier;
    private boolean initialState;
    private boolean finalState;
    private  LinkedList<TransducerEdge> edges;

    public TransducerNode(String id, Boolean initialState, Boolean finalState) {
        this.identifier = id;
        this.initialState = initialState;
        this.finalState = finalState;
        this.edges = new LinkedList<>();
    }

    public boolean isInitialState() {
        return initialState;
    }

    public boolean isFinalState() {
        return finalState;
    }

    public String getIdentifier() {
        return identifier;
    }

    public LinkedList<TransducerEdge> getEdges() {
        return edges;
    }

    public boolean equals(TransducerNode otherNode) {
        // compare basic parameters
        if (compareToExcludingEdges(otherNode)) {
            // compare edges
            {
                if (this.edges == null || otherNode.edges == null) {
                    return false;
                }

                if (this.edges.size() != otherNode.edges.size()) {
                    return false;
                }

                return equalsOtherEdgeSet(otherNode.edges);
            }
        }

        return false;
    }

    private boolean equalsOtherEdgeSet(LinkedList<TransducerEdge> otherEdgeList) {
        int pairsFound = 0;
        for (TransducerEdge thisEdge : this.edges) {
            for (TransducerEdge otherEdge : otherEdgeList) {
                if (thisEdge.equals(otherEdge)) {
                    pairsFound++;
                    break;
                }
            }
        }
        return pairsFound == (this.edges.size());
    }

    public boolean compareToExcludingEdges(TransducerNode otherNode) {
        return this.initialState == otherNode.initialState && this.finalState == otherNode.finalState && this.identifier.equals(otherNode.identifier);
    }
}

File: tin/src/main/kotlin/tin/model/transducer/TransducerNode.kt
package tin.model.transducer

import tin.model.compareEdgeSets
import java.util.*

class TransducerNode(
        val identifier: String,
        val isInitialState: Boolean,
        val isFinalState: Boolean,
        val edges: LinkedList<TransducerEdge>?
) {

    constructor(identifier: String,
                isInitialState: Boolean,
                isFinalState: Boolean
    ) : this(
            identifier = identifier,
            isInitialState = isInitialState,
            isFinalState = isFinalState,
            edges = LinkedList())

    fun equals(otherNode: TransducerNode): Boolean {
        // compare basic parameters
        if (compareToExcludingEdges(otherNode)) {
            // compare edges
            run {
                if (this.edges == null || otherNode.edges == null) {
                    return false
                }
                return if (this.edges.size != otherNode.edges.size) {
                    false
                } else compareEdgeSets(this.edges, otherNode.edges)
            }
        }
        return false
    }


    fun compareToExcludingEdges(otherNode: TransducerNode): Boolean {
        return isInitialState == otherNode.isInitialState && isFinalState == otherNode.isFinalState && identifier == otherNode.identifier
    }
}
